.github/workflows/unit_tests.sh

```diff
@@ -27,6 +27,25 @@ function info() {
 
 set -ex
 
+export PATH="$HOME/.cargo/bin:$HOME/.local/bin:$PATH"
+
+function install_rust() {
+    local tmp=$(mktemp -d)
+    # The ciphers and url are borrowed from https://sh.rustup.rs
+    local ciphers=TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384
+    local url="https://static.rust-lang.org/rustup/dist/x86_64-unknown-linux-gnu"
+
+    curl -V | grep -q ' OpenSSL/'
+    curl --proto =https --tlsv1.2 --ciphers $ciphers --silent --show-error --fail --location "$url/rustup-init" --output "$tmp/rustup-init"
+    curl --proto =https --tlsv1.2 --ciphers $ciphers --silent --show-error --fail --location "$url/rustup-init.sha256" --output "$tmp/rustup-init.sha256"
+    cmp <(sha256sum $tmp/rustup-init | awk '{ print $1 }') <(awk '{ print $1 }' "$tmp/rustup-init.sha256")
+    chmod +x "$tmp/rustup-init"
+    "$tmp/rustup-init" -y --default-toolchain nightly
```

Comment:
Is there any reason you are using nightly? Most distros only ship stable AFAIK. Systemd is also not a kernel where you may need to compile the standard library yourself for the respective target with options to for example disable sse usage or need other nightly features.

Reply:
I think it was for ASAN support - only the nightly has it?

Reply:
Ah, I see. Probably a good idea to test on stable too to prevent depending on unstable features or features that have only recently been stabilized.

Reply:
Yeah the "build_tests" job does that, it just installs rustc from the distro. This one ("unit_tests") needs to run with ASAN so this workaround was added. It would be great if it wasn't necessary - with GCC/Clang you don't need nightly versions of the compiler suites to run with sanitizers.

Reply:
I see. Rust code being memory safe by default means that you don't need ASAN most of the time. And miri can detect most of the bugs in unsafe code that ASAN and UBSAN detect and also certain kinds of UB that can't be detected by generic sanitizers as they are Rust specific rules. Because of this I think there isn't much pressure into stabilizing ASAN support, but I may be wrong here. Miri isn't suitable for this particular use case though as it is an interpreter and thus much much slower and doesn't have C interoperability. Only a hard coded list of libc functions are emulated.

Reply:
Given the `unsafe` wrappers usage, it will be very important to run the tests under ASAN for all involved components.

Reply:
I agree. Wherever there is `unsafe` you should run at least some kind of sanitizer. ASAN is likely the best in this case.

Reply:
> Yeah the "build_tests" job does that, it just installs rustc from the distro

It just so happens that `rustc` on GHActions is new enough to be compatible with all the `clang` versions `systemd` is built with there. It's just a matter of time before a few stable rust releases should be maintained there to avoid clang/rustc mismatches (caused by LTO mostly).

> Is there any reason you are using nightly?

Apart from ASan, another reason to use nightly is to use the -Z instrument-coverage compiler flag

Speaking of coverage reports, @bjorn3 I wonder if there are projects written in both C and Rust I can take a look at where coverage reports are generated properly? Just to clarify, I'm trying to figure out how to get https://oss-fuzz.com/coverage-report/job/libfuzzer_asan_systemd/latest to work with C and Rust.

Reply:
@bjorn3 Thanks for mentioning Miri! Looks like that's exactly what I was looking for.

> Miri isn't suitable for this particular use case though as it is an interpreter and thus much much slower and doesn't have C interoperability

It seems it isn't suitable as long as the Rust code piggybacks on the C testing infrastructure only. The string functions can be covered by Miri as far as I can see (at least I managed to run `string_extract_line` with `Miri` and even triggered a strategically placed stack-buffer-overflow):
```
error: Undefined Behavior: memory access failed: pointer must be in-bounds at offset 20, but is outside bounds of alloc1700 which has size 16
  --> src/main.rs:26:14
   |
26 |     let _y = *xs.as_ptr().offset(4);
   |              ^^^^^^^^^^^^^^^^^^^^^^ memory access failed: pointer must be in-bounds at offset 20, but is outside bounds of alloc1700 which has size 16
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
   = note: inside `string_extract_line` at src/main.rs:26:14
```

The question is how to replace `cargo miri` with `meson` and `RUSTFLAGS` and keep it more or less maintainable.